sha2/src/sha256/aarch64.rs

```diff
@@ -1,15 +1,134 @@
 //! SHA-256 `aarch64` backend.
 
+// Implementation adapted from mbedtls.
+
 // TODO: stdarch intrinsics: RustCrypto/hashes#257
 
+use core::arch::aarch64::*;
+use core::arch::asm;
+
+use crate::consts::K32;
+
 cpufeatures::new!(sha2_hwcap, "sha2");
 
 pub fn compress(state: &mut [u32; 8], blocks: &[[u8; 64]]) {
     // TODO: Replace with https://github.com/rust-lang/rfcs/pull/2725
     // after stabilization
     if sha2_hwcap::get() {
-        sha2_asm::compress256(state, blocks);
+        for block in blocks {
+            unsafe { sha256_compress(state, block) }
+        }
     } else {
         super::soft::compress(state, blocks);
     }
 }
+
+macro_rules! vsha256hq_u32 {
```

Comment:
IME, Rust can’t fully inline functions in the same way it inlines intrinsics. It introduces some value copying which I haven’t been able to eliminate and which can have negative effects on performance relative to the macro approach. I suspect it’s due to `asm!` being a bit of a black box for the compiler but I’m a bit out of my depth there.

“This works gooder on my laptop” is probably the more concise explanation, though.

Reply:
Hmm yeah, was just playing with the AES intrinsics in godbolt and the ASM it's generating with a function wrapper isn't quite as good as what invoking the intrinsic does, instead there are some ancillary `ldp`/`stp`/`mov` instructions.

Haven't tried seeing what it does in context with `#[inline(always)]` though... it may be able to elide them in that case?

Reply:
I wasn’t able to get rid of the copying with `#[inline(always)]` in my AEGIS-128L impl, so I don’t think it’d work here.

Reply:
Wow, you weren't kidding. I just got `aes` working with polyfill functions and it's about 30% slower versus the intrinsics, even with `#[inline(always)]`

Reply:
Interesting! Maybe I'm doing something else wrong then...

Reply:
I just tried it and in this context there doesn’t seem to be a penalty, whereas for AES it’s stiff. I wonder if that’s due to the `AESE`/`AESMC` fusion being interrupted?

I’ll change these macros to functions.

Reply:
@codahale that could definitely be it, yeah. Might just need to have a block of inline assembly in the core encryption/decryption functions with both instructions for now

Reply:
Ok, they’re now just functions and the performance is exactly the same.